### PR TITLE
Add model behavior analysis utility under kornia.contrib Closes #3626

### LINKED ISSUE #3626: [Feature]: Model Behavior Analysis under Kornia Augmentations (kornia.contrib)
### 🚀 Feature Description

Kornia provides powerful differentiable augmentations, but there is currently no built-in tooling to analyze how vision models behave internally when exposed to these augmentations.

I propose adding a small diagnostic utility that applies Kornia augmentations to an input image and analyzes how intermediate model representations change across layers. The goal is to measure layer sensitivity and representation stability under augmentation-induced perturbations.

The feature would primarily target research and debugging workflows and could live under kornia.contrib as experimental tooling.

### 📂 Feature Category

Augmentation

### 💡 Motivation

When training or evaluating vision models with augmentations, it is often unclear how those perturbations affect internal representations of the network. If model performance changes under augmentations, it is difficult to determine which layers are sensitive to the perturbations and whether the learned representations remain stable.

Currently, Kornia provides powerful augmentation operators, but there is no direct way within the ecosystem to inspect how models respond internally to those transformations.

Providing a lightweight analysis utility could help users better understand representation robustness, debug model behavior under augmentations, and study invariance properties of vision models.

### 💭 Proposed Solution

One possible approach would be to implement a small analysis utility that applies Kornia augmentations to an input image and measures how model representations change under those perturbations.

The workflow could look roughly as follows:

image
↓
Kornia augmentation pipeline
↓
model forward pass
↓
capture intermediate layer activations
↓
compute layer sensitivity scores
↓
analyze representation stability for selected layers
↓
evaluate robustness of the final embedding

Intermediate activations could be captured using PyTorch forward hooks, allowing the tool to remain model-agnostic and compatible with common PyTorch vision architectures.

Since this functionality is primarily intended for analysis and experimentation rather than core vision operators, it could potentially live under kornia.contrib.

### 🔄 Alternatives Considered

The same type of analysis can technically be performed using ad-hoc scripts by applying augmentations and manually inspecting intermediate activations. However, this approach usually requires implementing forward hooks, feature comparison logic, and analysis pipelines from scratch for each experiment.

While there are external robustness and interpretability libraries, they are generally designed around their own augmentation frameworks and do not integrate naturally with Kornia’s differentiable augmentation pipeline.

Providing a small utility within the Kornia ecosystem would make this type of representation analysis easier to perform while keeping the workflow fully compatible with Kornia-based augmentation pipelines.

### 🎯 Use Cases

• Robustness analysis – Evaluate how sensitive different layers of a vision model are to common augmentations such as rotations, blur, or brightness changes.

• Model debugging – Identify layers where representations change significantly under small perturbations, which can help diagnose instability or brittleness in trained models.

• Representation analysis – Study how internal feature representations evolve across layers when inputs are perturbed using Kornia augmentations.

• Research workflows – Support experiments related to invariance, robustness, and representation stability in vision models using Kornia-based augmentation pipelines.

### 📝 Additional Context

The analysis would rely on PyTorch forward hooks to capture intermediate activations and would remain model-agnostic so it can work with common PyTorch vision architectures.

The augmentation stage would leverage existing Kornia augmentation pipelines (e.g., AugmentationSequential), allowing users to analyze model behavior under the same transformations used during training or evaluation.

Since this functionality is primarily intended for experimentation and diagnostics rather than core operators, it would likely fit best under kornia.contrib.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Aditya-ha11**
This PR introduces a utility to analyze how vision model internal representations
change under Kornia augmentations.

This utility helps diagnose model robustness by quantifying representation shifts across layers.

## Key features
- Applies augmentations to input images  
- Extracts intermediate representations using forward hooks  
- Computes layer-wise differences between original and augmented inputs  
- Supports multiple layers  

## Use cases
- Understanding augmentation sensitivity  
- Debugging model robustness  
- Research on representation stability  

## Basic example
```python
analyze_model_behavior(model, image, augmentation, layers=("layer1", "layer2"))
```

## Usage example
```python
from kornia.contrib import analyze_model_behavior
import kornia.augmentation as K

aug = K.RandomGaussianBlur((3, 3))

results = analyze_model_behavior(
    model,
    image,
    augmentation=aug,
    layers=("layer1", "layer3")
)

print(results)
```

## Example output
```python
{
  "layer1": 0.12,
  "layer3": 0.67
}
```

## Interpretation of results:

- Low difference (0.0 – 0.2):
  The model is robust to the augmentation at this layer.

- Moderate difference (0.3 – 0.6):
  The model shows partial sensitivity.

- High difference (> 0.6):
  The model is highly sensitive and representations change significantly.

## Actionable insights:

- High difference in early layers:
  Indicates sensitivity to low-level features.
  → Action: Improve data augmentation or preprocessing.

- High difference in deeper layers:
  Indicates unstable semantic representations.
  → Action: Fine-tune the model or improve training data.

- Consistently low differences across layers:
  Indicates strong robustness.
  → Action: Suitable for deployment in real-world noisy settings.

- Inconsistent differences across layers:
  Indicates representation instability.
  → Action: Investigate architecture or apply regularization.

### CODE REVIEW

**Comment by navya7821** [comment]
Hey, I opened the original proposal (#3626) and had mentioned my intent to implement this.

I’ve been working on a more extended version of this idea as a standalone framework (covering feature extraction, augmentation pipelines, and multi-metric analysis like CKA, similarity, and robustness scoring), so I won’t be contributing to this PR.

I’ll share my implementation separately once it’s ready.

**Comment by shijianjian** [comment]
Hi @Aditya-ha11, I like the idea. Can you please add some examples on how to use it for users? Essentially, what kind of knowledge can we extract from your analysis results? Then, according to the results, what actions shall one take?

**Comment by Aditya-ha11** [comment]
Hi! Thanks for the helpful suggestion.

I’ve updated the PR with:

* A complete usage example and sample output
* Clear interpretation of the layer-wise differences
* Actionable insights on how users can respond to the results

Please let me know if there’s anything else you’d like improved!

**Comment by shijianjian** [comment]
Those numbers, such as {"layer1": 0.12}, depend on the model you are using, right? If you are using EfficientNet, its numerical scale might be smaller; you will be getting numbers such as {"layer1": 0.0000012}. So, I think you should give a schema to measure a percentage (against the layer-wise gradient) or something? Or should it be getting from multiple runs? It sounds like a research project to me. Are there any references about this?

Meanwhile, if this tool is an analyzer, I would prefer it to generate a report directly, by giving a model directly with a default config, rather than manually inputting layers and operations.

@navya7821 Do you mean the same as this PR? I thought you were mentioning the differentiability? Can you provide any reference to what you are planning to contribute, including those metrics you mentioned, and how to use those metrics? 

In general, we are happy to work on some research projects, but I do not think this PR is mature enough to be merged.

**Comment by navya7821** [comment]
> Those numbers, such as {"layer1": 0.12}, depend on the model you are using, right? If you are using EfficientNet, its numerical scale might be smaller; you will be getting numbers such as {"layer1": 0.0000012}. So, I think you should give a schema to measure a percentage (against the layer-wise gradient) or something? Or should it be getting from multiple runs? It sounds like a research project to me. Are there any references about this?
> 
> Meanwhile, if this tool is an analyzer, I would prefer it to generate a report directly, by giving a model directly with a default config, rather than manually inputting layers and operations.
> 
> @navya7821 Do you mean the same as this PR? I thought you were mentioning the differentiability? Can you provide any reference to what you are planning to contribute, including those metrics you mentioned, and how to use those metrics?
> 
> In general, we are happy to work on some research projects, but I do not think this PR is mature enough to be merged.

Thanks for the discussion here — this is quite insightful.

I don’t mean exactly the same scope as this PR. My proposal is aimed at a more general **representation-level analysis under augmentations**, rather than focusing on a single metric.

For context, I had opened a broader discussion earlier:
https://github.com/kornia/kornia/issues/3626

To make this more concrete, I’ve been developing a **standalone prototype**:
**https://github.com/navya7821/representation-profiling-framework**

The current design focuses on:


**1. layer-wise sensitivity profiling across augmentations
2. a model-agnostic analysis pipeline
3. an extensible structure for adding metrics like feature drift / stability
4. reproducible, config-driven analysis (same setup yields consistent results)
5. and automatic report-style analysis from a model + config (no manual layer selection)**

In particular, it already produces a **consolidated report-style output** summarizing behavior across layers (e.g., sensitivity scores, stability metrics, and robustness indicators), which aligns with the idea of making this more of an analyzer rather than a manual metric tool.

Regarding metrics, I agree that raw values can be scale-dependent — the direction I’m exploring is toward normalized / relative measures (e.g., percentage changes or distribution-based comparisons across layers).

If this direction makes sense, I can work on narrowing this into smaller, well-scoped components that fit naturally into Kornia (e.g., core hooks + one concrete metric).This is currently a baseline implementation to validate the core design, and I plan to refine and extend it further based on feedback.

Here’s a sample output from the current prototype:
image

**Comment by Aditya-ha11** [comment]
> Those numbers, such as {"layer1": 0.12}, depend on the model you are using, right? If you are using EfficientNet, its numerical scale might be smaller; you will be getting numbers such as {"layer1": 0.0000012}. So, I think you should give a schema to measure a percentage (against the layer-wise gradient) or something? Or should it be getting from multiple runs? It sounds like a research project to me. Are there any references about this?
> 
> Meanwhile, if this tool is an analyzer, I would prefer it to generate a report directly, by giving a model directly with a default config, rather than manually inputting layers and operations.
> 
> @navya7821 Do you mean the same as this PR? I thought you were mentioning the differentiability? Can you provide any reference to what you are planning to contribute, including those metrics you mentioned, and how to use those metrics?
> 
> In general, we are happy to work on some research projects, but I do not think this PR is mature enough to be merged.

Hi! Thanks a lot for the thoughtful feedback — really helpful.

You’re right about the scale issue. I’ve updated the implementation to normalize the differences using the magnitude of the original activations, so the values are now more consistent and comparable across different models.

I also did some quick checks with real images and stronger augmentations to make sure the results are meaningful — for example, deeper layers tend to show higher sensitivity, which is what we’d expect.

On the usability side, I agree it would be nicer if users didn’t have to manually specify layers. I’ll look into adding a simple default setup for that.

For now, I’ve kept things lightweight, but this could definitely be extended later (e.g., multiple runs or generating a small report).

Let me know if this direction makes sense or if you’d like me to adjust anything further!
